src/d3d9/d3d9_shader_validator.h

```diff
@@ -37,32 +71,203 @@ namespace dxvk {
+        return ErrorCallback(pFile, Line, 0,
+          D3D9ShaderValidatorMessage::InstructionNullArgs,
+          "IDirect3DShaderValidator9::Instruction called with NULL == pdwInst or 0 == cdw.");
+      }
+
+      if (unlikely(m_state == D3D9ShaderValidatorState::Begin)) {
+        return ErrorCallback(pFile, Line, 0,
+          D3D9ShaderValidatorMessage::InstructionOutOfOrder,
+          "IDirect3DShaderValidator9::Instruction called out of order. ::Begin must be called first.");
+      } else if (unlikely(m_state == D3D9ShaderValidatorState::EndOfShader)) {
+        return ErrorCallback(pFile, Line, 0,
+          D3D9ShaderValidatorMessage::InstructionEndOfShader,
+          "IDirect3DShaderValidator9::Instruction called out of order. After end token there should be no more instructions. Call ::End next.");
+      } else if (unlikely(m_state == D3D9ShaderValidatorState::Error)) {
+        return E_FAIL;
+      }
+
+      if (m_state == D3D9ShaderValidatorState::ValidatingHeader)
+        return ValidateHeader(pFile, Line, pdwInst, cdw);
+
+      DxsoCodeIter pdwInstIter{ reinterpret_cast<const uint32_t*>(pdwInst) };
+      bool isEndToken = !m_ctx->decodeInstruction(pdwInstIter);
+      const DxsoInstructionContext instContext = m_ctx->getInstructionContext();
+
+      if (isEndToken)
+        return ValidateEndToken(pFile, Line, pdwInst, cdw);
+
+      // TODO: DxsoDecodeContext::decodeInstructionLength() does not currently appear
+      // to return the correct token length in many cases, and as such dwordLength
+      // will not be equal to cdw in many situations that are expected to pass validation
+      //
+      /*Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: opcode ", instContext.instruction.opcode));
+      // + 1 to account for the opcode...
+      uint32_t dwordLength = instContext.instruction.tokenLength + 1;
+      Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: cdw ", cdw));
+      Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: dwordLength ", dwordLength));
+      if (unlikely(cdw != dwordLength)) {
+        return ErrorCallback(pFile, Line, 0x2,
+          D3D9ShaderValidatorMessage::BadInstructionLength,
+          str::format("Instruction length specified for instruction (", cdw, ") does not match the token count encountered (", dwordLength, "). Aborting validation."));
+      }*/
+
+      // a maximum of 10 inputs are supported with PS 3.0 (validation required by The Void)
+      if (m_isPixelShader && m_majorVersion == 3) {
+        switch (instContext.instruction.opcode) {
+          case DxsoOpcode::Comment:
+          case DxsoOpcode::Def:
+          case DxsoOpcode::DefB:
+          case DxsoOpcode::DefI:
+            break;
+
+          default:
+            for (uint32_t instNum = 1; pdwInst[instNum] & 0x80000000; instNum++) {
+              DWORD regType  = ((pdwInst[instNum] & D3DSP_REGTYPE_MASK)  >> D3DSP_REGTYPE_SHIFT)
```

Comment:
Is it possible to also get those from `DxsoInstructionContext`?

Reply:
Will check.

Reply:
Might be problematic because `DxsoInstructionContext` gets the register token count from `decodeInstructionLength` and that might be different than iterating over all tokens that have bit 31 set.
And broken shaders are why we're implementing this after all.

Reply:
Ugh, I see... decodeInstructionLength does not appear to work correctly in some cases in its current implementation, as per the commented section in the above validation that was added by @misyltoad . That's a bit beyond my scope to fix at this point.

Reply:
We *could* replace the for loop by iterating over source registers decoded by `decodeInstruction` and `decodeInstructionLength` but I guess just leave it as it is for now before we make this more complicated than it has to be.